src/js/mediaStore.js

```diff
@@ -0,0 +1,108 @@
+  };
+
+  // In a real version, the initial state of this could likely just be a generic transform
+  // of the `stateFacade`
+  let state = Object.freeze({
+    mediaPaused: stateFacade.mediaPaused.get(stateOwners),
+  });
+
+  if (stateOwners.media) {
+    // This is similar to the `<media-controller>` setup for state monitoring
+    // and state propagation
+    Object.entries(stateFacade).forEach(([stateName, { get, mediaEvents }]) => {
+      // In a full version, we'd want to store these somewhere so
+      // we can remove if the media is removed/replaced
+      const handler = () => {
+        const nextValue = get(stateOwners);
+        if (state[stateName] === nextValue) return;
+
+        // Update the state since it changed.
+        // Using an "immutable" approach here so
+        // callbacks can easily do comparisons between prev/next state.
+        // Freezing isn't necessary, though it's a light touch enforcement
+        // of immutability (in case folks try to directly modify state)
+        state = Object.freeze({
+          ...state,
+          [stateName]: nextValue,
+        });
+
+        // Given anything that cares the updated state
+        callbacks.forEach((cb) => cb(state));
+      };
+      mediaEvents.forEach((eventType) => {
+        stateOwners.media.addEventListener(eventType, handler);
+      });
+    });
+  }
+
+  return {
+    // note these "actions" could just be the events if we wanted, especially if we normalize on "detail" for
+    // any payload-relevant values
+    dispatch({ type, detail }) {
+      if (
+        [
+          MediaUIEvents.MEDIA_PAUSE_REQUEST,
+          MediaUIEvents.MEDIA_PLAY_REQUEST,
+        ].includes(type)
+      ) {
+        // Most if not all state change requests can be modeled as UPDATE requests (in the generic "CRUD" sense)
+        // 1. Name of state to change (in this case, "paused")
+        // 2. Value you want the state to be (in this case, true if MEDIA_PAUSE_REQUEST, false if MEDIA_PLAY_REQUEST)
+        // We can think of more specific requests (e.g. MEDIA_PAUSE_REQUEST, MEDIA_PLAY_REQUEST) as just shorthand
+        // translations of e.g. { type: 'mediaupdaterequest', key: nameOfState, value: valueOfState }. In this case,
+        // 'mediapauserequest'|'mediaplayrequest' can be thought of as shorthand/convenience actions for something like:
+        // { type: 'mediaupdaterequest', detail: { key: 'mediaPaused', value: true | false } }
+        const key = 'mediaPaused';
+        const value = type === MediaUIEvents.MEDIA_PAUSE_REQUEST;
+        stateFacade[key].set(value, stateOwners);
+      }
+      // if (type === 'mediaelementchangerequest') {} // we can also e.g. get updates for the stateOwners themselves
+    },
+    getState() {
+      // return the current state, whatever it is
+      return state;
+    },
+    subscribe(callback) {
+      // whenever state has (potentially?) changed based on "state  owners events", invoke callback(latestState)
```

Comment:
So every state receiver would get every state change with this, right? I understand this is a pattern, and I don't necessarily assume optimization makes a noticeable difference here, but this is less optimal then the current implementation where state receivers are more selective of what state they want to be updated with, right?

Reply:
Yes-ish and definitely worth calling out/thinking through. Typically the updates are mitigated through a "selector" pattern that identifies the relevant state and effectively memoizes whether or not some effect will occur only if/when a "change" for the "interesting state" happens. I'm doing a toy example of this in the higher order callback used for subscribe in `<media-controller>`